File: ios/AgStatus/SSEClient.swift
//
//  SSEClient.swift
//  AgStatus
//
//  Server-Sent Events client for `<board>/events`. Streams bytes via
//  URLSession, parses SSE frames incrementally (event:/data: lines
//  terminated by a blank line, ':' comment lines ignored) and yields
//  typed events. The store owns reconnect/backoff.
//

import Foundation

// MARK: - SSEEvent

enum SSEEvent: Sendable {
    case snapshot([Session])
    case upsert(Session)
    case remove(id: String)
}

// MARK: - SSEClient

final class SSEClient: Sendable {

    private let session: URLSession

    init() {
        let configuration = URLSessionConfiguration.ephemeral
        // The server sends a keepalive comment every 25s, so a healthy
        // stream never gets close to this request timeout.
        configuration.timeoutIntervalForRequest = 75
        // The stream itself is expected to stay open indefinitely.
        configuration.timeoutIntervalForResource = 365 * 24 * 3600
        configuration.requestCachePolicy = .reloadIgnoringLocalCacheData
        session = URLSession(configuration: configuration)
    }

    /// Connects to `<board>/events` and yields parsed events until the
    /// connection drops (stream finishes) or fails (stream throws:
    /// 404 → APIError.boardNotFound, 429 → .rateLimited, network → .unreachable).
    func events(for board: Board) -> AsyncThrowingStream<SSEEvent, Error> {
        let session = self.session
        let url = board.boardURL.appendingPathComponent("events")

        return AsyncThrowingStream { continuation in
            let task = Task {
                var request = URLRequest(url: url)
                request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
                request.cachePolicy = .reloadIgnoringLocalCacheData

                do {
                    try AgStatusAPI.checkTransportSecurity(url)
                    let (bytes, response) = try await session.bytes(for: request)
                    guard let http = response as? HTTPURLResponse else {
                        throw APIError.unreachable
                    }
                    // Check status BEFORE consuming the stream.
                    switch http.statusCode {
                    case 200: break
                    case 404: throw APIError.boardNotFound
                    case 429: throw APIError.rateLimited
                    case 503: throw APIError.atCapacity
                    default: throw APIError.badResponse(http.statusCode)
                    }

                    // Parse raw bytes on newline boundaries. (AsyncLineSequence
                    // is avoided on purpose: it can skip the blank lines that
                    // delimit SSE frames.)
                    var parser = SSEParser()
                    for try await byte in bytes {
                        try Task.checkCancellation()
                        if let event = parser.consume(byte) {
                            continuation.yield(event)
                        }
                    }
                    continuation.finish()
                } catch let error as APIError {
                    continuation.finish(throwing: error)
                } catch is CancellationError {
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: APIError.unreachable)
                }
            }
            continuation.onTermination = { _ in
                task.cancel()
            }
        }
    }
}

// MARK: - SSEParser

/// Incremental SSE wire-format parser. Feed it bytes; it returns a decoded
/// event whenever a complete frame (terminated by a blank line) arrives.
private struct SSEParser {
    private var lineBuffer: [UInt8] = []
    private var eventName: String?
    private var dataLines: [String] = []

    mutating func consume(_ byte: UInt8) -> SSEEvent? {
        guard byte == 0x0A else { // \n
            lineBuffer.append(byte)
            return nil
        }
        if lineBuffer.last == 0x0D { // strip trailing \r
            lineBuffer.removeLast()
        }
        let line = String(decoding: lineBuffer, as: UTF8.self)
        lineBuffer.removeAll(keepingCapacity: true)
        return process(line)
    }

    private mutating func process(_ line: String) -> SSEEvent? {
        if line.isEmpty {
            return flush()
        }
        if line.hasPrefix(":") { // comment / keepalive
            return nil
        }

        let field: Substring
        var value: Substring
        if let colon = line.firstIndex(of: ":") {
            field = line[line.startIndex..<colon]
            value = line[line.index(after: colon)...]
            if value.hasPrefix(" ") { // spec: strip a single leading space
                value = value.dropFirst()
            }
        } else {
            field = line[...]
            value = ""
        }

        switch field {
        case "event":
            eventName = String(value)
        case "data":
            dataLines.append(String(value))
        default:
            break // id:, retry:, unknown fields — ignored
        }
        return nil
    }

    /// Builds an event from the accumulated frame, if any. Multi-line data
    /// is joined with "\n" per the SSE spec. Undecodable frames are dropped.
    private mutating func flush() -> SSEEvent? {
        defer {
            eventName = nil
            dataLines.removeAll(keepingCapacity: true)
        }
        guard !dataLines.isEmpty else { return nil }

        let payload = Data(dataLines.joined(separator: "\n").utf8)
        let decoder = JSONDecoder()

        switch eventName ?? "message" {
        case "snapshot":
            guard let sessions = try? decoder.decode([Session].self, from: payload) else { return nil }
            return .snapshot(sessions)
        case "session":
            guard let session = try? decoder.decode(Session.self, from: payload) else { return nil }
            return .upsert(session)
        case "remove":
            guard let removal = try? decoder.decode(RemovePayload.self, from: payload) else { return nil }
            return .remove(id: removal.id)
        default:
            return nil
        }
    }

    private struct RemovePayload: Decodable {
        let id: String
    }
}

File: android/app/src/main/kotlin/com/kardanov/agstatus/SseClient.kt
package com.kardanov.agstatus

import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.buffer
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import okhttp3.CacheControl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.sse.EventSource
import okhttp3.sse.EventSourceListener
import okhttp3.sse.EventSources
import java.util.concurrent.TimeUnit

/**
 * Server-Sent Events client for `<board>/events`. okhttp-sse parses the wire
 * format; this maps the named frames onto typed events. The store owns
 * reconnect and backoff.
 */

// MARK: - SseEvent

sealed interface SseEvent {
    data class Snapshot(val sessions: List<Session>) : SseEvent
    data class Upsert(val session: Session) : SseEvent
    data class Remove(val id: String) : SseEvent
    data class Usage(val usage: List<UsageInfo>) : SseEvent
}

// MARK: - SseClient

class SseClient {

    private val client = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        // The stream is expected to stay open indefinitely — the server sends a
        // keepalive comment every 25s — so read and call timeouts are disabled.
        .readTimeout(0, TimeUnit.MILLISECONDS)
        .callTimeout(0, TimeUnit.MILLISECONDS)
        .build()

    /**
     * Connects to `<board>/events` and emits parsed events until the stream
     * ends (flow completes) or fails (flow throws: 404 → boardNotFound,
     * 429 → rateLimited, network → unreachable). Unknown event names and
     * undecodable payloads are dropped, never fatal.
     */
    fun events(board: Board): Flow<SseEvent> = callbackFlow {
        val url = board.boardUrl.toHttpUrlOrNull()?.newBuilder()?.addPathSegment("events")?.build()
            ?: throw ApiException.unreachable()
        val request = Request.Builder()
            .url(url)
            .header("Accept", "text/event-stream")
            .cacheControl(CacheControl.FORCE_NETWORK)
            .build()

        val listener = object : EventSourceListener() {
            override fun onEvent(eventSource: EventSource, id: String?, type: String?, data: String) {
                decodeEvent(type, data)?.let { trySend(it) }
            }

            override fun onClosed(eventSource: EventSource) {
                close()
            }

            override fun onFailure(eventSource: EventSource, t: Throwable?, response: Response?) {
                val status = response?.code
                close(
                    if (status != null && status !in 200..299) ApiException.forStatus(status)
                    else ApiException.unreachable()
                )
            }
        }

        val source = EventSources.createFactory(client).newEventSource(request, listener)
        awaitClose { source.cancel() }
    }
        // The store's state is a fold over this stream, so a slow collector must
        // never cost an event.
        .buffer(Channel.UNLIMITED)

    private fun decodeEvent(type: String?, data: String): SseEvent? = when (type) {
        "snapshot" -> decodeOrNull<List<Session>>(data)?.let { SseEvent.Snapshot(it) }
        "session" -> decodeOrNull<Session>(data)?.let { SseEvent.Upsert(it) }
        "remove" -> decodeOrNull<RemovePayload>(data)?.let { SseEvent.Remove(it.id) }
        "usage" -> decodeOrNull<List<UsageInfo>>(data)?.let { SseEvent.Usage(it) }
        else -> null
    }

    private inline fun <reified T> decodeOrNull(data: String): T? =
        try {
            AgStatusJson.decodeFromString<T>(data)
        } catch (malformed: IllegalArgumentException) {
            null
        }

    @Serializable
    private data class RemovePayload(val id: String)
}
